### PR TITLE
Complete action for the TMOP Integrator

### PR DESCRIPTION
**Pull request by kmittal2**
This PR adds the terms that were missing from the action of the TMOP Integrator. This impacts cases with discrete and analytic adaptivity where the targets were a function of the position in the domain.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1567](https://github.com/mfem/mfem/pull/1567) | @kmittal2 | @tzanio | @vladotomov + @tzanio | 06/27/20 | 07/16/20 | 07/21/20 | |

### CODE REVIEW

File: fem/tmop.cpp [hunk 1]
```diff
@@ -970,6 +982,45 @@ void AnalyticAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
+{
+   DenseMatrix point_mat;
+   point_mat.UseExternalData(elfun.GetData(), fe.GetDof(), fe.GetDim());
+
+   switch (target_type)
+   {
+      case GIVEN_FULL:
+      {
+         MFEM_VERIFY(matrix_tspec != NULL,
+                     "Target type GIVEN_FULL requires a MatrixCoefficient.");
```
Comment (vladotomov) [inline fem/tmop.cpp:1000]: TMOPMatrixCoefficient

File: fem/tmop.cpp [hunk 2]
```diff
@@ -970,6 +982,45 @@ void AnalyticAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
+
+         IsoparametricTransformation Tpr;
+         Tpr.SetFE(&fe);
+         Tpr.ElementNo = e_id;
+         Tpr.GetPointMat().Transpose(point_mat);
```
Comment (vladotomov) [inline fem/tmop.cpp:1005]: Seems that the function expects `Tpr` as input, no need to reconstruct it.

File: fem/tmop.hpp [hunk 3]
```diff
@@ -677,6 +677,44 @@ class TargetConstructor
...
+   /** @note When this method is called, the caller must make sure that the
+       IntegrationPoint associated with @a T is the same as @a ip. This can be
+       achieved by calling T.SetIntPoint(&ip). */
+   virtual void Eval(DenseMatrix &K, ElementTransformation &T,
+                     const IntegrationPoint &ip) = 0;
+
+   /** @brief Evaluate the derivative of the matrix coefficient with respect
+       to @a comp in the element described by @a T at the point @a ip,
+       storing the result in @a K. */
+   virtual void EvalP(DenseMatrix &K, ElementTransformation &T,
```
Comment (vladotomov) [inline fem/tmop.hpp:714]: How about `EvalGrad`?

File: fem/tmop.hpp [hunk 4]
```diff
@@ -677,6 +677,44 @@ class TargetConstructor
...
+
+   TMOPMatrixCoefficient(int h, int w) : MatrixCoefficient(h, w)  { }
+
+   void SetTime(double t) { time = t; }
+   double GetTime() { return time; }
+
+   int GetHeight() const { return height; }
+   int GetWidth() const { return width; }
+   // For backward compatibility
+   int GetVDim() const { return width; }
```
Comment (vladotomov) [inline fem/tmop.hpp:701]: These function don't need to be here, we only need the first constructor and the Evals.

File: fem/tmop.cpp [hunk 5]
```diff
@@ -1767,13 +2177,44 @@ void TMOP_Integrator::AssembleElementVectorExact(const FiniteElement &el,
...
+         // dWinv/dx = -Winv*dW/dx*Winv
+         MultAtB(PMatI, DSh, Amat);
+         for (int d = 0; d < dim; d++)
+         {
+            const DenseMatrix &dJtr_q = dJtr(q + d*nqp);
+            Mult(Jrt, dJtr_q, work1); //Winv*dw/dx
+            Mult(work1, Jrt, work2);  //Winv*dw/dx*Winv
+            Mult(Amat, work2, work1); //A*Winv*dw/dx*Winv
+            MultAtB(P, work1, work2); //dmu/dT^T*A*Winv*dw/dx*Winv
+            d_Winv_dx(d) = work2.Trace(); //Tr[dmu/dT : AWinv*dw/dx*Winv]
```
Comment (vladotomov) [inline fem/tmop.cpp:2207]: Is the `Trace` needed here? It's not in the comment before the loop.

File: fem/tmop.cpp [hunk 6]
```diff
@@ -1368,6 +1422,353 @@ void DiscreteAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
    }
 }
 
+void DiscreteAdaptTC::ComputeElementTargetsGradient(int e_id,
+                                                    const FiniteElement &fe,
+                                                    const IntegrationRule &ir,
```
Comment (vladotomov) [inline fem/tmop.cpp:1427]: `e_id` and `fe` are part of `Tpr`.

File: fem/tmop.cpp [hunk 7]
```diff
@@ -1368,6 +1422,353 @@ void DiscreteAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
+
+   switch (target_type)
+   {
+      case IDEAL_SHAPE_GIVEN_SIZE:
+      case GIVEN_SHAPE_AND_SIZE:
+      {
+         const DenseMatrix &Wideal =
+            Geometries.GetGeomToPerfGeomJac(fe.GetGeomType());
+         const int dim = Wideal.Height(),
+                   ndofs = tspec_fes->GetFE(0)->GetDof(),
```
Comment (vladotomov) [inline fem/tmop.cpp:1446]: Won't work for mixed meshes :)

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
LGTM

**Pull request merged**